### DISCUSSION TITLE
watch() and getValues() not returning file object for input type file

### DISCUSSION THREAD

**Discussion by sujeet-agrahari** [react-hook-form/react-hook-form#8208] [Q&A]
I have a component which uploads a file. When an image is selected it previews it in another component. I am using material UI library.

So, when an image is selected,  I need to access it. I tried using `getValues()` and `watch()` from `react-hook-form` but it is returning me path of the file in string form not the file object.

```
          <div className="photo">
            <Avatar
              src={
                // it gives path of the file as string
                getValues('photo')
                  ? URL.createObjectURL(getValues('photo')[0])
                  : "https://icon-library.com/images/no-image-icon/no-image-icon-0.jpg"
              }
              alt="photo preview"
              sx={{ width: "200px", height: "200px" }}
            />
            <label htmlFor="contained-button-file">
              <Controller
                name="photo"
                control={control}
                render={({ field }) => (
                  <Input
                    {...field}
                    accept="image/*"
                    id="contained-button-file"
                    type="file"
                    error={Boolean(errors["photo"])}
                    helperText={errors["photo"]?.message}
                  />
                )}
              />
              <Button variant="contained" component="span">
                Upload
              </Button>
            </label>
          </div>
```

How to access selected file object?

**Comment by Moshyfawn** [comment]
That's the default behavior of the native Web input of type "file".

Check your `Input`'s API and see if it returns files along with the `event` object. If not, you can access your files like so:
```js
onChange={event => event.target.files} 
```

Keep in mind that native Web inputs don't take any other value but a path strong of the file. Seeing that you're using a controlled input, you will have to manage passing the file name back into the input's value and storing the actual files

**Reply by sujeet-agrahari** [reply]
>That's the default behavior of the native Web input of type "file".

What behaviour?

It does return the file when I register `onChange` event. Why it doesn't do so with` watch()` or `getValues()`? 
See this example https://codesandbox.io/s/react-hook-form-uploadfile-with-thumbnail-preview--drag-and-drop-p11gy?file=/src/FileInput.jsx

It is returning files with `watch(name)`.

You can also see here https://dev.to/coder4_life/upload-files-with-react-hook-form-m4j , he is doing the same.

**Reply by Moshyfawn** [reply]
> What behaviour?

From [MDN Web docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#value):

> Value | A DOMString representing the path to the selected         file.
> -- | --

When you pass the event into your field's `onChange` handler, the form value is set to the file path of the file you've uploaded. Use `event.target.files` to store your files in RHF instead.

Also, to "subscribe" to the value change, use `watch` instead of `getValues`. See [RHF docs](https://react-hook-form.com/api/useform/watch) to learn more.

I'd recommend to omit watching the file input value all together and just collect all your `Avatar` and `Input` components under one `Controller` for simplicity and some performance gains.

[Here's a working CSB example](https://codesandbox.io/s/objective-wildflower-856lkb?file=%2Fsrc%2FDemo.tsx)

**Reply by sujeet-agrahari** [reply]
Thanks. I had make it worked using `setValue()` :(

Using work around for now. Using [setValue()][1] and calling it on change of the input file.

```
          <div className="photo">
            <Avatar
              src={
                watch('photo')
                  ? URL.createObjectURL(watch('photo')[0])
                  : "https://icon-library.com/images/no-image-icon/no-image-icon-0.jpg"
              }
              alt="photo preview"
              sx={{ width: "200px", height: "200px" }}
            />
            <label htmlFor="contained-button-file">
              <Input
                accept="image/*"
                id="contained-button-file"
                type="file"
                error={Boolean(errors["photo"])}
                onChange={(e) => setValue("photo", e.target.files)}
                multiple
                helperText={errors["photo"]?.message}
              />

              <Button variant="contained" component="span">
                Upload
              </Button>
            </label>
          </div>
```
But, you are absolutely right, thanks again. 

  [1]: https://react-hook-form.com/api/useform/setvalue

**Reply by Moshyfawn** [reply]
What do you need to use `setValue`?

**Reply by ritikbanger** [reply]
Instead of using RHF for images, use custom functions, it would be a better approach.

 ```
  /**
   * This ref is used to take files input.
   */

  const inputRef: React.RefObject<HTMLInputElement> = React.createRef()
  const onEdit = () => {
    inputRef.current?.click()
  }                      
                       <Box onClick={onEdit}>
                        <input
                        type='file'
                        onChange={(e) => handleFileUpload(e)}
                        accept='image/jpg, image/jpeg, image/png'
                        multiple
                        hidden
                        ref={inputRef}
                      />
                     </Box>
```

```
 const [imageErrors, setImageErrors] = useState<Interface>()
  const [imagesInfo, setImagesInfo] = useState<Interface>({
    imagesUrl: [],
    file: [],
  })

  /**
   * This function is used to handle images and set images in a state.
   */
  const handleFileUpload = async ({ currentTarget: { files } }: React.ChangeEvent<HTMLInputElement>) => {
    setImageErrors({})
    if (files && files.length !== 0) {
      const filesArray = Array.from(files)
      const sizeErrorFiles: string[] = []
      const formatErrorFiles: string[] = []
      const array = await filesArray.filter((file) => {
        if (file.size < 2 * 1024 * 1024) {
          if (!file.type.match(validationConstants.imageRegex)) {
            formatErrorFiles.push(file.name)
          } else {
            return file
          }
        } else {
          sizeErrorFiles.push(file.name)
        }

        return null
      })
      setImageErrors({ imageFormatError: formatErrorFiles, imageSizeError: sizeErrorFiles })
      if (array) {
        setImagesInfo({
          imagesUrl: [
            ...imagesInfo.imagesUrl,
            ...array.map((file) => ({
              url: URL.createObjectURL(file),
            })),
          ],
          file: [...imagesInfo.file, ...array],
        })
      }
    }
  }

```

**Reply by sujeet-agrahari** [reply]
> What do you need to use `setValue`?

I was using this to set the value of the `photo` to the `e.target.files` so that It can be accessed on data when form is submitted.

**Reply by Moshyfawn** [reply]
> > What do you need to use `setValue`?
> 
> I was using this to set the value of the `photo` to the `e.target.files` so that It can be accessed on data when form is submitted.

You can simply use an object to store your image `path` and the `file` and update both fields `onChange`.

See [this updated CSB](https://codesandbox.io/s/objective-wildflower-856lkb?file=%2Fsrc%2FDemo.tsx)

```jsx
<Controller
  name="avatar"
  control={control}
  render={({
    field: {
      value: { path, file },
      onChange,
      ...fieldProps
    }
  }) => (
    <Stack gap={2}>
      <Avatar
        sx={{ width: 90, height: 90 }}
        src={file ? URL.createObjectURL(file) : placeholderImgUrl}
        alt="avatar preview"
      />
      <label htmlFor="avatar">
        <Input
          {...fieldProps}
          value={path}
          onChange={(event) =>
            onChange({
              path: event.target.value,
              file: event.target.files[0]
            })
          }
          type="file"
          accept="image/*"
          id="avatar"
        />
        <Button variant="contained" component="span">
          Upload
        </Button>
      </label>
    </Stack>
  )}
/>
```
